### ISSUE TITLE
How to show a context menu on right-click on a TreeView item?

### ISSUE THREAD

**Issue by ScottHutchinson** [elmish/Elmish.WPF#517]
What would be the best binding to use to respond to a right-click on a TreeView item?

I added a MouseRightButtonUp binding at https://github.com/ScottHutchinson/StructFilters/blob/main/NG-DART%20StructFilters.Views/StructFiltersWindow.xaml#L47

You can see my first attempt to add a field binding at https://github.com/ScottHutchinson/StructFilters/blob/main/NG-DART%20StructFilters/StructFilters.fs#L262

That is sort of working. It gives me the context (the item that the user clicked on), but unfortunately, it is firing many times and not when expected. I need it to fire only once, so I can show a context menu and perform the selected action on the item that was clicked on.

Thanks!

**Additional Information**
Elmish.WPF version: 3.5.6
Target framework version: net472

**Comment by LyndonGingerich** [comment]
To respond to a user action, you should use `Binding.cmd`, which sends a message from the view. `Binding.cmd` tells the model when the view wants a change; `Binding.oneWay` tells the view when the model changes.

**Comment by ScottHutchinson** [comment]
@LyndonGingerich Thanks. That makes sense. But how to code that to figure out which TreeView element the user clicked on?

**Comment by LyndonGingerich** [comment]
Right. If you have a collection of items with their own bindings, you want to use `Binding.subModelSeq`. This will allow you to set up bindings for each item individually.

**Comment by LyndonGingerich** [comment]
Using the `subModelSeq` binding for `ItemsControl.ItemsSource`, you can call individual bindings (from the `bindings` argument of `Binding.subModelSeq`) from within `ItemsControl.ItemTemplate`.

**Comment by ScottHutchinson** [comment]
I am already using `Binding.subModelSeq`, but I don't know how to integrate a `Binding.cmd` with that. All the `Binding.cmd` instances I have so far are in the rootBindings.

**Comment by LyndonGingerich** [comment]
Ooh, nice structure. I've been wanting to learn about rose trees.

Can you put a `cmd` binding that returns a `RoseTreeMsg` in `fieldBindings`?

**Comment by ScottHutchinson** [comment]
Can you put a cmd binding that returns a RoseTreeMsg in fieldBindings?

That might be what I should do, but so far I don't know how to do that.

**Comment by LyndonGingerich** [comment]
`Binding.cmd` is overloaded. It doesn't have to take just the message you want to return; it can also take a `'model -> 'msg`. So you can use `Binding.cmd` in `fieldBindings` with an argument of type `Model * (RoseTree<FieldData> * RoseTree<FieldData>) -> RoseTreeMsg<FieldId, SubtreeMsg>`.

**Comment by LyndonGingerich** [comment]
[Here](https://github.com/elmish/Elmish.WPF/blob/master/TUTORIAL.md#command-bindings) is the overload reference in case it helps later.

**Comment by ScottHutchinson** [comment]
Without a concrete example, I'm stuck. the closest I've found is https://github.com/elmish/Elmish.WPF/blob/a3dc5678c09cc22e02ffcb56ec9516263897e88b/src/Samples/NewWindow.Core/Window2.fs#L70

**Comment by LyndonGingerich** [comment]
That looks quite plausible at first glance. Are you encountering a compilation error or incorrect behavior?

**Comment by LyndonGingerich** [comment]
For concrete examples, I'd probably check the [samples](https://github.com/elmish/Elmish.WPF/tree/master/src/Samples), or possibly [dependent projects](https://github.com/elmish/Elmish.WPF/network/dependents). But I find that the aforementioned tutorial usually solves my issues most quickly.

**Comment by ScottHutchinson** [comment]
I think I might have it working like this:

```
        "RightClick" |> Binding.cmd(fun (_, (_, (fd: RoseTree<FieldData>))) -> showContextMenu fd.Data)
```
where
```
    let private showContextMenu (fld: FieldData) : RoseTreeMsg<FieldId, SubtreeMsg> =
        // TODO: Show context menu and perform selected action here.
        LeafMsg RightClick
```
@LyndonGingerich Thank you for your help!

**Comment by TysonMN** [comment]
Thanks very much @LyndonGingerich for helping with this!

> Ooh, nice structure. I've been wanting to learn about rose trees.

A rose tree is a tree where the number of children varies, which means the children are stored in a list.  That is the kind of tree used in the `SubModelSeq` sample.

@ScottHutchinson, please let me know if you have any further questions about this.

**Comment by ScottHutchinson** [comment]
@TysonMN So, only half of the problem is solved. Showing a context menu is "solved", but the other half is handling the context menu items with the node that was right-clicked. I am trying to move the ContextMenu from the Window.Resources into the TreeView.ItemTemplate, but I'm not sure if that will work. For one thing, I'm not sure if FindResource can find the ContextMenu inside an ItemTemplate.

**Comment by LyndonGingerich** [comment]
> I'm attempting to add this line

That line is blank. Is the link correct?

**Comment by ScottHutchinson** [comment]
The link is correct. I meant adding the line below at that blank location.

```
let designVm = ViewModel.designInstance App.update (App.rootBindings())
```

**Comment by LyndonGingerich** [comment]
Right, my bad. Shouldn't `ViewModel.designInstance` take a sample of the model type (a `Domain.Model` in this case) instead of `App.update`?

**Comment by ScottHutchinson** [comment]
Maybe, but the compiler isn't complaining about `App.update`. It's complaining about `App.rootBindings()`, which returns `Binding<Model, Msg> list`. Not sure how to make the compiler happy.

**Comment by LyndonGingerich** [comment]
Are you sure it's not complaining about `App.rootBindings` because of the type inferred from `App.update`? I don't think it should be expecting a `Binding<(App.Msg -> Domain.Model -> Domain.Model),'a> list`.

**Comment by ScottHutchinson** [comment]
To use `designVm` above, I need to add the XAML code like below. But I think that will not work because the NG-DART_StructFilters F# project has a reference to the C# project (where the XAML is), not vice versa. Possibly I could reverse that relationship, but that could be a lot of work, so maybe there is an easier way to solve this problem.

```
xmlns:vm="clr-namespace:NGDartStructFilters;assembly=NG-DART_StructFilters"
```
...
```
d:DataContext="{x:Static vm:StructFilters.designVm}"
```

**Comment by LyndonGingerich** [comment]
I have never before considered referencing the view from the model rather than the other way around. If you're not sure that's what you want to do, then I wouldn't. If you are sure, I kind of want to see how it turns out. That's beyond my level of expertise.

**Comment by ScottHutchinson** [comment]
When I started working on this project in 2020, all the samples were like my project (F# model referencing the C# view). At some point they changed all the samples the other way around, but I left mine as it was. Maybe being able to access the ViewModel in the XAML is why they switched it like that.

**Comment by LyndonGingerich** [comment]
Interesting! You've been working with it longer than I have, then. I had assumed that Elmish.WPF was designed for onion or hexagonal architecture, moving I/O toward the outside.

**Comment by TysonMN** [comment]
> Maybe being able to access the ViewModel in the XAML is why they switched it like that.

Exactly.

**Comment by ScottHutchinson** [comment]
I don't understand this code that appears to pass a type `MainWindow` instead of an object. When I try to do that in my project, it does not compile. What trick makes that work?

At https://github.com/elmish/Elmish.WPF/blob/8367849a1a52665cc8d39fe53f46a78bcd8bbeac/README.md#L162

```
Program.main(MainWindow);
```

**Comment by ScottHutchinson** [comment]
Also, my C# project is a library, not an .exe, because in our solution the startup project is a C++ project.

In a library project, maybe I can't define an Application class like the samples, so I might need to figure out a different way to define the `StartElmish` function.

**Comment by TysonMN** [comment]
I will have a look as soon as I can. I am busy with work at the moment.

You have provided lik6e to code. Is that your actual code (including the C++ executable b project) or a minimal reproduction?

**Comment by ScottHutchinson** [comment]
I have provided only the two Elmish-related projects, not the C++ consumers.

**Comment by TysonMN** [comment]
I cloned and compiled your code, but I am not sure what else I should do there.

Can you share a minimal working example of your problem (or current behavior)?  If not, how about sharing a video (or gif) of the current behavior?

**Comment by ScottHutchinson** [comment]
I have modified my projects to match the samples (C# project referencing the F# project now).

But the binding is still not working. The breakpoint at https://github.com/ScottHutchinson/StructFilters/blob/main/NG-DART%20StructFilters/StructFilters.fs#L257 is still not hit when the user clicks the context menu item.

The diagnostics log says "DataContext is null".

**Comment by LyndonGingerich** [comment]
> the samples target net5.0

(net6.0 now, actually.)

**Comment by ScottHutchinson** [comment]
Should be net6.0, since net5.0 is no longer supported. But this one is still net5.0. https://github.com/elmish/Elmish.WPF/blob/8367849a1a52665cc8d39fe53f46a78bcd8bbeac/src/Samples/SubModelSeq.Core/SubModelSeq.Core.fsproj#L8

**Comment by ScottHutchinson** [comment]
> Pardon. I see the use of `Binding ContextSelectAll`, but where is `contextMenu` or `contextSelectAll` used?

https://github.com/ScottHutchinson/StructFilters/blob/main/NG-DART%20StructFilters/StructFilters.fs#L212

**Comment by ScottHutchinson** [comment]
> Can you share a minimal working example

I might have to do that, if I can't get it working on my own.

**Comment by LyndonGingerich** [comment]
Oh, I see. It's net6.0 on `master`, but not on that version.

**Comment by ScottHutchinson** [comment]
> Can you share a minimal working example of your problem

@TysonMN I replicated what I am trying to do in a clone/branch of the subModelSeq sample. To reproduce the problem, start the subModelSeq sample, add children to a node, then right-click on the parent node that has children, select the "Select All" context menu item, and observe that the breakpoint is not hit at this line: https://github.com/ScottHutchinson/Elmish.WPF/blob/master/src/Samples/SubModelSeq.Core/Program.fs#L248

**Comment by TysonMN** [comment]
Great work! I will investigate tonight.

**Comment by TysonMN** [comment]
~I think I made a lot of progress, but I didn't finish. Here is what I have so far in case it takes me a while to get back to this. The next time I expect to work on this is Thursday evening.~

I was able to get this working now (because I woke up early).

My approach explicitly puts the `ContextMenu` in the `HierarchicalDataTemplate` (and on the `StackPanel`) and then adds bindings to the corresponding `MenuItems`s.

Your approach explicitly puts the `ContextMenu` in the `Resources` of the `Window` and then uses `window.FindResource("contextMenu")` to get that `ContextMenu`.  In that case, I don't know what the `DataContext` is for the `ContextMenu`.  If it would have been the `DataContext` of the `Window`, then I think you would have been able to hit your desired breakpoint.  If you wanted to make more progress on your approach, then I would try setting the `DataContext` of the `ContextMenu` to be the same as the `Window`.  (I didn't try that though.)

However, even if you had the `DataContext` you wanted, I think it is the wrong `DataContext`.  If `Select All` is clicked, then we don't know which children should be selected.  We want the `DataContext` to the the `DataContext` of the element that was right clicked so we can get the IDs of its children ID for selection.

I didn't try to implement that selection.  Try to do that and let me know if you get stuck.  If you do, come ready with a minimal working example like you did here.

As a "hint", I expect the IDs to be (de)selected need to be passed up to the root so they can be added to (or removed from) a list of multi-selected IDs.  This code already uses the type `InOutMsg`.  IIRC, I changed the type of this type in my code at work to just `InOut`.  Then it makes more (semantic) sense to use this as a return type of `updateSubtree`, which can return
```fsharp
InOut<RoseTree<Identifiable<Openable<Counter>>>, Guid list>
```
 
https://github.com/TysonMN/Elmish.WPF/tree/people/ScottHutchinson/select_children

**Comment by ScottHutchinson** [comment]
@TysonMN, Thank you so much! I was able to run your sample, which worked as I expected. It is unfortunate though how intrusive this `Openable` type is. It changes the type of the Model, which means that much of the code must be changed to fix about 36 compile errors. I think I will try to come up with a less intrusive approach that uses everything you did, but without the `Openable` type.

**Comment by ScottHutchinson** [comment]
> I was able to run your sample, which worked as I expected.

@TysonMN , Actually it didn't work exactly as expected, since the first time I right-click on a parent node, nothing happens. The second time I right-click, then the context menu is shown.

**Comment by ScottHutchinson** [comment]
> the first time I right-click on a parent node, nothing happens.

I have everything else working in my application, except the user must right-click twice to show the context menu the first time they right-click a parent node.

**Comment by TysonMN** [comment]
Can you share a reproduction with that behavior?

**Comment by TysonMN** [comment]
Oh, I didn't notice that. I will take a look later tonight.

**Comment by TysonMN** [comment]
I fixed the problem.  I pushed [this commit](https://github.com/TysonMN/Elmish.WPF/commit/7791706ed2b3766b5e224c7bb184c20f8a7196cf) to that branch.

The issue has something to do with the `DataContext` of the `ContextMenu` being wrong.  But I don't understand in what way it was wrong or why.  During the first right click, a message containing `SetContextMenuIsOpen false` is dispatched.  That is wrong and explains why the context menu did not appear that time.  Subsequent right clicks caused `SetContextMenuIsOpen true` to be dispatched, which explains why the context menu did appear those times. 

My best guess is that the `DataContext` of the `ContextMenu` was initially `null`, so explicitly setting the `DataContext` (via a binding) ensures it has the correct initial value.

I got the exact code from [this page](https://www.codeproject.com/Articles/162784/WPF-ContextMenu-Strikes-Again-DataContext-Not-Upda).  However, some of the things they say there are no true.  For example, this statement is not true.

> You need this spell only if you expect parent's data context to change during the life of the parent.

The `DataContext` of the parent is not changing in our case, so this would imply that this code should change the behavior for us, but it does.

Whatever.  The problem is solved.  I would prefer a better understanding of the problem and solution, but I will have to be satisfied with this.
